subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/execution/PluginsBlockInterpreter.kt

```diff
@@ -71,50 +78,61 @@ fun interpret(program: Program.Plugins): PluginsBlockInterpretation = KotlinLexe
 
             BLOCK_COMMENT -> Unit
             DOC_COMMENT -> Unit
-
-            in WHITE_SPACE_OR_COMMENT_BIT_SET -> {
-                // detect newline separators between statements
-                if (state == InterpreterState.AFTER_ID && tokenSequence.contains('\n')) {
-                    newStatement()
-                }
-            }
+            EOL_COMMENT -> seenStatementSeparator = true
+            in WHITESPACES -> if (tokenSequence.contains('\n')) seenStatementSeparator = true
 
             else -> {
                 when (state) {
+
                     InterpreterState.START -> state = when (tokenType) {
                         LBRACE -> InterpreterState.ID
                         else -> return expecting("{")
                     }
 
                     InterpreterState.ID -> state = when (tokenType) {
                         RBRACE -> InterpreterState.END
-                        SEMICOLON -> InterpreterState.ID
-                        IDENTIFIER -> when (tokenText) {
-                            "id" -> InterpreterState.ID_OPEN_CALL
-                            "kotlin" -> InterpreterState.KOTLIN_ID_OPEN_CALL
-                            else -> return expecting("id or kotlin")
+                        SEMICOLON -> InterpreterState.ID.also { if (pluginId != null) newPluginRequest() }
+                        IDENTIFIER -> {
+                            if (pluginId != null && seenStatementSeparator) newPluginRequest()
+                            when (tokenText) {
+                                "id" -> InterpreterState.ID_OPEN_CALL
+                                "kotlin" -> InterpreterState.KOTLIN_ID_OPEN_CALL
+                                else -> return expecting("id or kotlin")
+                            }
                         }
 
                         else -> return expecting("plugin spec")
                     }
 
-                    InterpreterState.AFTER_ID -> {
-                        state = when (tokenType) {
-                            DOT -> state
-                            SEMICOLON -> newStatement()
-                            RBRACE -> {
-                                newStatement()
-                                InterpreterState.END
-                            }
+                    InterpreterState.AFTER_ID -> state = when (tokenType) {
+                        DOT -> {
+                            if (previousTokenType == DOT) return expecting("version or apply")
+                            InterpreterState.AFTER_ID
+                        }
 
-                            IDENTIFIER -> when (tokenText) {
+                        SEMICOLON -> newPluginRequest()
+                        RBRACE -> InterpreterState.END.also { newPluginRequest() }
+                        IDENTIFIER -> {
+                            when (tokenText) {
                                 "version" -> InterpreterState.VERSION_OPEN_CALL
```

Comment:
🧪 One more:

```kotlin
    @Test
    fun `syntax error - id() dot version`() {
        assertDynamicInterpretationOf(
            """id("plugin-id").version "1.0"""",
            """Expecting '(', got "1.0""""
        )
    }
```

By the way, I'm starting to consider that it might pay off to move to a combinator based setup or at least introduce a tokenizer to simplify the parsing state machine.

Reply:
Thanks, fixed on https://github.com/gradle/gradle/pull/22435

Yes I agree that the state machine will get out of hands if we add more features to it :+1: 

Especially if we consider implementing support for version catalog aliases, but I think we should not for now. I explained my rationale in https://github.com/gradle/gradle/pull/22435 description.

Moving to combinators could be done as a later step given we now have descent code coverage.